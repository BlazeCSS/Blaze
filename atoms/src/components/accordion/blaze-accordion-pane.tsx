import { Component, Method, Prop, State } from '@stencil/core';

@Component({
  tag: 'blaze-accordion-pane'
})
export class AccordionPane {

  @State() _isOpen: boolean = false;
  @Prop() open: boolean = false;
  @Prop() header: string;

  componentWillLoad() {
    this._isOpen = this.open;
  }

  @Method()
  show() {
    this._isOpen = true;
  }

  @Method()
  close() {
    this._isOpen = false;
  }

  toggle() {
    this._isOpen ? this.close() : this.show();
  }

  @Method()
  isOpen() {
    return this._isOpen;
  }

  render() {
    const isOpenClass = this._isOpen ? 'c-card__item--active' : '';

    return ([
      <div class={`c-card__item c-card__item--title ${isOpenClass}`} onClick={() => this.toggle()}>
        {this.header}
      </div>,
      <div class="c-card__item c-card__item--pane">
        <slot />
      </div>
    ]);
  }
}
