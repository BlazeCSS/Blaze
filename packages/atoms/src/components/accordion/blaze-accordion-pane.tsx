import { Component, Method, Prop, State } from '@stencil/core';

@Component({
  tag: 'blaze-accordion-pane'
})
export class AccordionPane {

  @State() _isOpen: boolean = false;
  @Prop() open: boolean = false;
  @Prop() header: string;

  componentWillLoad() {
    this._isOpen = this.open;
  }

  @Method()
  show() {
    this._isOpen = true;
  }

  @Method()
  close() {
    this._isOpen = false;
  }

  toggle() {
    this._isOpen ? this.close() : this.show();
  }

  @Method()
  isOpen() {
    return this._isOpen;
  }

  render() {
    const isOpenClass = this._isOpen ? 'c-card__control--active' : '';

    return ([
      <button
        role="heading"
        aria-expanded={this._isOpen}
        class={`c-card__control ${isOpenClass}`}
        onClick={() => this.toggle()}>

        {this.header}

      </button>,
      <section aria-hidden={!this._isOpen} class="c-card__item c-card__item--pane">
        <slot />
      </section>
    ]);
  }
}
