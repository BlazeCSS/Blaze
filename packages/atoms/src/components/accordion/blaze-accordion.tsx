import { Component } from '@stencil/core';

@Component({
  tag: 'blaze-accordion'
})
export class Accordion {

  render() {
    return (
      <div class="c-card c-card--accordion">
        <slot />
      </div>
    );
  }
}
