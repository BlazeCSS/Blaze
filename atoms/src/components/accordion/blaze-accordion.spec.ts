import { flush, render } from '@stencil/core/testing';
import { Accordion } from './blaze-accordion';
import { AccordionPane } from './blaze-accordion-pane';

describe('Accordion', () => {
  it('should build', () => {
    expect(new Accordion()).toBeTruthy();
    expect(new AccordionPane()).toBeTruthy();
  });

  describe('renders', () => {
    let element;

    const snapIt = (name, html) => {
      it(name, async () => {
        element = await render({
          components: [Accordion, AccordionPane],
          html
        });
        await flush(element);

        expect(element).toMatchSnapshot();
      });
    };

    snapIt(
      'an emtpy accordion',
      `<blaze-accordion></blaze-accordion>`
    );

    snapIt(
      'two headers and two panes',
      `<blaze-accordion>
        <blaze-accordion-pane header="Click me">
          This is an accordion
        </blaze-accordion-pane>
        <blaze-accordion-pane header="Click me also">
          To toggle other panes
          </blaze-accordion-pane>
      </blaze-accordion>`
    );
  });

  it('headers open panes', async () => {
    const element = await render({
      components: [Accordion, AccordionPane],
      html: `<blaze-accordion>
              <blaze-accordion-pane header="Click me">
                This is an accordion
              </blaze-accordion-pane>
              <blaze-accordion-pane header="Click me also">
                To toggle other panes
                </blaze-accordion-pane>
            </blaze-accordion>`
    });
    await flush(element);

    element.querySelector('blaze-accordion-pane').show();
    await flush(element);

    expect(element).toMatchSnapshot();
  });

  it('headers close panes', async () => {
    const element = await render({
      components: [Accordion, AccordionPane],
      html: `<blaze-accordion>
              <blaze-accordion-pane header="Click me">
                This is an accordion
              </blaze-accordion-pane>
              <blaze-accordion-pane header="Click me also">
                To toggle other panes
                </blaze-accordion-pane>
            </blaze-accordion>`
    });
    await flush(element);

    element.querySelector('blaze-accordion-pane').show();
    await flush(element);

    element.querySelector('blaze-accordion-pane').close();
    await flush(element);

    expect(element).toMatchSnapshot();
  });

  it('clicking a header opens the pane', async () => {
    const element = await render({
      components: [Accordion, AccordionPane],
      html: `<blaze-accordion>
              <blaze-accordion-pane header="Click me">
                This is an accordion
              </blaze-accordion-pane>
              <blaze-accordion-pane header="Click me also">
                To toggle other panes
                </blaze-accordion-pane>
            </blaze-accordion>`
    });
    await flush(element);

    expect(element.querySelector('blaze-accordion-pane').isOpen()).toEqual(false);


    element.querySelector('blaze-accordion-pane div').click();
    await flush(element);

    expect(element.querySelector('blaze-accordion-pane').isOpen()).toEqual(true);
  });
});
