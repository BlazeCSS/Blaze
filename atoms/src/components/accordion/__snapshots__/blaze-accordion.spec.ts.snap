// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion headers close panes 1`] = `
<blaze-accordion
  class="hydrated"
>
  <div
    class="c-text c-card c-card--accordion"
  >
    <!---->
    
              
    <blaze-accordion-pane
      class="hydrated"
      header="Click me"
    >
      <div
        class="c-card__item c-card__item--title"
      >
        Click me
      </div>
      <div
        class="c-card__item c-card__item--pane"
      >
        <!---->
        
                This is an accordion
              
      </div>
    </blaze-accordion-pane>
    
              
    <blaze-accordion-pane
      class="hydrated"
      header="Click me also"
    >
      <div
        class="c-card__item c-card__item--title"
      >
        Click me also
      </div>
      <div
        class="c-card__item c-card__item--pane"
      >
        <!---->
        
                To toggle other panes
                
      </div>
    </blaze-accordion-pane>
    
            
  </div>
</blaze-accordion>
`;

exports[`Accordion headers open panes 1`] = `
<blaze-accordion
  class="hydrated"
>
  <div
    class="c-text c-card c-card--accordion"
  >
    <!---->
    
              
    <blaze-accordion-pane
      class="hydrated"
      header="Click me"
    >
      <div
        class="c-card__item c-card__item--title c-card__item--active"
      >
        Click me
      </div>
      <div
        class="c-card__item c-card__item--pane"
      >
        <!---->
        
                This is an accordion
              
      </div>
    </blaze-accordion-pane>
    
              
    <blaze-accordion-pane
      class="hydrated"
      header="Click me also"
    >
      <div
        class="c-card__item c-card__item--title"
      >
        Click me also
      </div>
      <div
        class="c-card__item c-card__item--pane"
      >
        <!---->
        
                To toggle other panes
                
      </div>
    </blaze-accordion-pane>
    
            
  </div>
</blaze-accordion>
`;

exports[`Accordion renders an emtpy accordion 1`] = `
<blaze-accordion
  class="hydrated"
>
  <div
    class="c-text c-card c-card--accordion"
  >
    <!---->
  </div>
</blaze-accordion>
`;

exports[`Accordion renders two headers and two panes 1`] = `
<blaze-accordion
  class="hydrated"
>
  <div
    class="c-text c-card c-card--accordion"
  >
    <!---->
    
        
    <blaze-accordion-pane
      class="hydrated"
      header="Click me"
    >
      <div
        class="c-card__item c-card__item--title"
      >
        Click me
      </div>
      <div
        class="c-card__item c-card__item--pane"
      >
        <!---->
        
          This is an accordion
        
      </div>
    </blaze-accordion-pane>
    
        
    <blaze-accordion-pane
      class="hydrated"
      header="Click me also"
    >
      <div
        class="c-card__item c-card__item--title"
      >
        Click me also
      </div>
      <div
        class="c-card__item c-card__item--pane"
      >
        <!---->
        
          To toggle other panes
          
      </div>
    </blaze-accordion-pane>
    
      
  </div>
</blaze-accordion>
`;
